Qt: Only call tryGetBalances in pollBalanceChanged if the result will be used.

tecnovert opened the pull request:
getBalances takes significant processing if the wallet has many transactions.

Commit:
Avoid calling tryGetBalances in pollBalanceChanged if the result won't be used.
getBalances takes significant processing if the wallet has many transactions.

sipa commented:
Concept ACK, seems reasonable if it's an issue.

MarcoFalke commented:
Could do an early return to avoid wrapping the whole function scope in two
sets of braces and indenting it by 4 spaces

On Sat, Jul 7, 2018 at 2:33 PM, Pieter Wuille <notifications@github.com>
wrote:

> Concept ACK, seems reasonable if it's an issue.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/13612#issuecomment-403254845>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGGmv7wlo1WfWRJHWaMjiwiIGZ2r5nfSks5uEWFPgaJpZM4VGdUl>
> .
>

Commit:
Rearranged so TRY_LOCK is called first.
The LOCK in getNumBlocks shouldn't be run before the TRY_LOCK in
tryGetBalances
The neatest fix seems to be to move the height check into tryGetBalances.

tecnovert commented:
It takes a fair amount of blocks and transactions before the issue shows up.
CPU usage scales with both the chain height and number of transactions.

gprof results show most of the time is taken by all the GetDepthInMainChain calls from IsTrusted

Testing on a regtest node with 10k blocks and 10k txns htop shows:
+-6 CPU% before the changes and
+-0.7 CPU% after

20k blocks and 10k txns:
+-6.5 CPU% before the changes and
+-0.7 CPU% after

40k blocks and 39.6k txns:
+-20 CPU% before the changes and
+-0.7 CPU% after

promag commented:
Concept ACK.

A greater improvement would be to avoid `pollBalanceChanged` each 250ms.

practicalswift commented:
Concept ACK

Nice first-time contribution @tecnovert!

jonasschnelli commented:
The way how to GUI (and the wallet) calculates the balance has significant performance impacts. I tried in the past to enhance this situation (but failed, see #10251).
Concept ACK, will test and perf-compare soon.

meshcollider reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/13612/commits/899d8f68a09d57d579b81a7e1d721a51e8c611bf, still needs rebase though

jarolrod commented:
@fanquake This can have the `Up For Grabs` label removed. A similar fix was done in https://github.com/bitcoin/bitcoin/pull/18160, then reverted as part of a different fix in https://github.com/bitcoin/bitcoin/pull/18587
